encoding/protobuf.go

```diff
@@ -0,0 +1,96 @@
+package encoding
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/jhump/protoreflect/desc"
+	"github.com/jhump/protoreflect/dynamic"
+	"github.com/yarpc/yab/protobuf"
+	"github.com/yarpc/yab/transport"
+)
+
+type protoSerializer struct {
+	serviceName string
+	methodName  string
+	method      *desc.MethodDescriptor
+}
+
+// NewProtobuf returns a protobuf serializer.
+func NewProtobuf(fullMethodName string, source protobuf.ProtoDescriptorSource) (Serializer, error) {
+	svc, mth, err := splitMethod(fullMethodName)
+	if err != nil {
+		return nil, err
+	}
+	dsc, err := source.FindSymbol(svc)
+	if err != nil {
+		return nil, fmt.Errorf("failed to query for service for symbol %q: %v", svc, err)
+	}
+	sd, ok := dsc.(*desc.ServiceDescriptor)
+	if !ok {
+		return nil, fmt.Errorf("target server does not expose service %q", svc)
+	}
+	mtd := sd.FindMethodByName(mth)
+	if mtd == nil {
+		return nil, fmt.Errorf("service %q does not include a method named %q", svc, mth)
+	}
+	return &protoSerializer{
+		serviceName: svc,
+		methodName:  mth,
+		method:      mtd,
+	}, nil
+}
+
+func (p protoSerializer) Encoding() Encoding {
+	return Protobuf
+}
+
+func (p protoSerializer) Request(body []byte) (*transport.Request, error) {
+	req := dynamic.NewMessage(p.method.GetInputType())
+	if err := req.UnmarshalJSON(body); err != nil {
+		return nil, fmt.Errorf("could not parse given request body as message of type %q: %v", p.method.GetInputType().GetFullyQualifiedName(), err)
+	}
+	bytes, err := proto.Marshal(req)
+	if err != nil {
+		return nil, fmt.Errorf("could marshal message of type %q: %v", p.method.GetInputType().GetFullyQualifiedName(), err)
+	}
+	return &transport.Request{
+		Method: fmt.Sprintf("%s::%s", p.serviceName, p.methodName),
+		Body:   bytes,
+	}, nil
+}
+
+func (p protoSerializer) Response(body *transport.Response) (interface{}, error) {
+	resp := dynamic.NewMessage(p.method.GetOutputType())
+	if err := resp.Unmarshal(body.Body); err != nil {
+		return nil, fmt.Errorf("could not parse given response body as message of type %q: %v", p.method.GetInputType().GetFullyQualifiedName(), err)
+	}
+	str, err := resp.MarshalJSON()
+	if err != nil {
+		return nil, err
+	}
+	var objmap map[string]*json.RawMessage
+	if err = json.Unmarshal(str, &objmap); err != nil {
```

Comment:
Missed this, shouldn’t this be jsonpb? Encoding/protobuf and the protobuf json format are not compatible

Reply:
*encoding/json

Reply:
This is to transform a json string into a `map[string]interface{}`, so that yab can display the response on the cli.  The json string is marshalled from binary to the string by https://godoc.org/github.com/jhump/protoreflect/dynamic#Message.MarshalJSON), so I think in this case this is valid.

Could add a comment for clarity.

Reply:
Why just just print the string from MarshalIndent? The jsonpb Marshaler has indent options too. Protobuf encoded as json cant be validly marshalled or unmarshalled using encoding/json, there are edge cases. See the protobuf json spec for all the details.

Reply:
Because I don't need to return a string but an `interface{}` type that yab understands for printing to the user (a native golang map or slice).

I need `proto bytes` -> `go map`, which I'm currently doing `bytes` -> `string` with `jsonpb` and then `string` to `map[string]interface{}` with `encoding/json`. I'm not trying to unmarshal the bytes directly with `encoding/json`. Is there a way to unmarshal bytes into a go type without going through json (for which I think I need both jsonpb and encoding/json)?

Reply:
Unmarshalling jsonpb-marshalled strings using encoding/json has the potential to fail, it’ll work until it potentially doesn’t.

Is it difficult to just make yab understand strings to print instead of map or slice? That’s the best bet.

Reply:
I believe this solves the problem of incorporating the JSON interpretation of the Protobuf response body in the JSON envelope `{"response": …}`, which carries the headers and status code depending on the transport protocol. This is important to ensure that we can unpack yab output with `jq` or similar. I believe any JSON string generated by Protobuf should be parseable by `encoding/json`.